packages/grafana-runtime/src/utils/queryResponse.ts

```diff
@@ -24,56 +25,63 @@ interface DataResponse {
 /**
  * Parse the results from /api/ds/query into a DataQueryResponse
  *
+ * @param res - the HTTP response data.
+ * @param queries optional DataQuery array that will order the response based on the order of query refId's.
+ *
  * @public
  */
-export function toDataQueryResponse(res: any): DataQueryResponse {
+export function toDataQueryResponse(res: any, queries?: DataQuery[]): DataQueryResponse {
   const rsp: DataQueryResponse = { data: [], state: LoadingState.Done };
   if (res.data?.results) {
     const results: KeyValue = res.data.results;
-    for (const refId of Object.keys(results)) {
+    const refIDs = queries ? queries.map((q) => q.refId) : Object.keys(results);
```

Comment:
does the backend now guarantee that everything has a refId set?  If not, this makes me a bit nervous

Reply:
It depends on how you see it I guess. The tsdb response model has a results map of refID's with query result:
https://github.com/grafana/grafana/blob/1bdd3eb3dd096862bebf6af59d68ff344cf3693c/pkg/tsdb/models.go#L33-L36

There's nothing in the backend that guarantees that only RefID's part of the queries are part of the response. Very much   up to the developer I would say, but still this should be "best practices" and documented at least. It's not really a new thing that you must return the result with RefID's of the queries to make certain feature in Grafana UI function as expected, for example error handling and so on.

To minimize the risk we could check here if there's any other RefID's in the response and include those in the result of the function. WDYT?